plugins/server_logs/lib/samson_server_logs/kafka_listener.rb

```diff
@@ -0,0 +1,120 @@
+
+    def initialize(deploy_id, output)
+      @deploy_id = deploy_id
+      @output = output
+      @consumer = Poseidon::PartitionConsumer.new(
+        "samson_deploy_listener",
+        KAFKA_HOST, KAFKA_PORT,
+        KAFKA_TOPIC, 0, :latest_offset
+      )
+    end
+
+    def listen
+      config = {
+        timeout: TOTAL_TIMEOUT.from_now,
+        servers: nil,
+        print: false
+      }
+
+      kafka = listen_to_kafka(config)
+      server_listener = listen_for_server_list(config)
+
+      deploy_success = yield
+
+      # no servers found during deploy ?
+      unless config.fetch(:servers)
+        kafka.kill
+        server_listener.kill
+        write "did not find any servers in deploy log"
+        return false
+      end
+
+      # wait for this time after the deploy is done
+      config[:timeout] = AFTER_DEPLOY_TIMEOUT.from_now
+      config[:print] = :all
+
+      return false unless kafka.value
+
+      deploy_success
+    end
+
+    private
+
+    # listen for the line that tells us which servers we actually deploy to
+    def listen_for_server_list(config)
+      Thread.new do
+        @output.each do |_, message|
+          if servers = message[/uploading deploy logs: \S+ servers: (\[.*?\])/i, 1]
+            config[:servers] = JSON.load(servers)
+            break
+          end
+        end
+      end
+    end
+
+    def listen_to_kafka(config)
+      Thread.new do
+        messages = []
+        done_servers = []
+        loop do
+          received = listen_for_messages(messages)
+
+          print = decide_what_to_print(config, received, messages)
+          print.each { |m| write m }
+
+          done_servers.concat(print.map { |m| m["host"] if m["exit"] == 0 }) # TODO handle other exit codes ...
+
+          # stop when we timeout or all servers are done
+          if Time.now > config.fetch(:timeout)
+            write "listener timed out"
+            break
+          elsif waiting_for_servers = config.fetch(:servers)
+            done_servers.each { |s| waiting_for_servers.delete(s) }
+            done_servers.clear
+
+            if waiting_for_servers.none?
+              write "received messages from all hosts"
+              break
+            end
+          end
+        end
+
+        (config.fetch(:servers) || []).empty? # did we complete ?
+      end
+    end
+
+    def write(message)
+      @output.write("Kafka: #{message}\n")
+    end
+
+    # start printing after the deploy is done, and then print everything as it comes in
+    def decide_what_to_print(config, received, messages)
+      case config.fetch(:print)
+      when :all
+        config[:print] = :new
+        messages
+      when :new
+        received
+      else
+        []
+      end
+    end
+
+    # listen for all messages for the current deploy
+    def listen_for_messages(messages)
+      received = @consumer.fetch(max_wait_ms: 100).map!(&:value).map { |m| JSON.load(m) }
+      received.select! { |m| m["deploy"] == @deploy_id }
```

Comment:
We'll have to send it to the deploy script somehow ... but we need
something unique

On Thu, Apr 30, 2015 at 10:28 AM, Steven Davidovitz <
notifications@github.com> wrote:

> In plugins/server_logs/lib/samson_server_logs/kafka_listener.rb
> <https://github.com/zendesk/samson/pull/379#discussion_r29450567>:
>
> > +    def decide_what_to_print(config, received, messages)
> > +      case config.fetch(:print)
> > +      when :all
> > +        config[:print] = :new
> > +        messages
> > +      when :new
> > +        received
> > +      else
> > +        []
> > +      end
> > +    end
> > +
> > +    # listen for all messages for the current deploy
> > +    def listen_for_messages(messages)
> > +      received = @consumer.fetch(max_wait_ms: 100).map!(&:value).map { |m| JSON.load(m) }
> > +      received.select! { |m| m["deploy"] == @deploy_id }
>
> I'm not 100% sure where this id will come from
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zendesk/samson/pull/379/files#r29450567>.
>